runtime/src/main/cpp/Memory.cpp

```diff
@@ -1825,12 +1792,12 @@ OBJ_GETTER(swapHeapRefLocked,
     shallRelease = oldValue != nullptr;
   }
   unlock(spinlock);
-  if (shallRelease) {
-    releaseHeapRef(oldValue);
-  }
   // No need to rememberNewContainer(), as oldValue is already
   // present on this worker.
```

Comment:
Btw, is this correct? Doesn't seem so when `oldValue != expectedValue`.

Reply:
Comment was indeed wrong.

Reply:
Comment is still wrong, and the behaviour is wrong too.
The following test crashes:
```kotlin
import kotlin.test.*
import kotlin.native.concurrent.*

val atomicRef = AtomicReference<Any?>(Any().freeze())
val semaphore = AtomicInt(0)

@Test fun runTest() {
    semaphore.value = 0
    val worker = Worker.start()
    val future = worker.execute(TransferMode.SAFE, { null }) {
        val value = atomicRef.compareAndSwap(null, null)
        semaphore.increment()
        while (semaphore.value != 2) {}
        println(value.toString() != "")
    }
    while (semaphore.value != 1) {}
    atomicRef.value = null
    kotlin.native.internal.GC.collect()
    semaphore.increment()
    future.result
    worker.requestTermination().result
}
```
(based on `worker10`).
Generally `atomicRef.compareAndSwap(Any().freeze(), null)` is the same as `atomicRef.value`, and if the latter requires `rememberNewContainer`, then the former requires this too.

Reply:
Thanks for bringing my attention to this problem. Fixed.